planseqlearn/psl/metaworld_mp_env.py

```diff
@@ -173,154 +176,269 @@ def get_body_geom_ids_from_robot_bodies(self):
-                        self, 49, ["corner", "corner2"], 500, 500, self.sim
-                    )
-                else:
-                    placement_pos = get_pose_from_sam_segmentation(
-                        self, object_name, "corner"
-                    )
-                placement_pos += np.array([0.13, 0.0, 0.15])
-        return placement_pos, placement_quat
-
-    def get_placement_poses(self):
-        return [self.get_placement_pose()]
+                camera_to_world = np.linalg.inv(world_to_camera)
+                object_pixels = np.argwhere(object_mask)
+                object_pointcloud = CU.transform_from_pixels_to_world(
+                    pixels=object_pixels,
+                    depth_map=depth_map[..., 0],
+                    camera_to_world_transform=camera_to_world,
+                )
+                self.sam_object_pose[obj_name] = np.mean(object_pointcloud, axis=0) + np.array([0.0, 0.0, 0.06])
+            if "hammer" in obj_name:
+                frame = self.sim.render(camera_name="corner3", width=500, height=500)
+                obj_masks, _, _, pred_phrases, _ = get_seg_mask(
+                    frame[:, :, ::-1],
+                    self.dino,
+                    self.sam,
+                    text_prompts=["robot, small red hammer handle"],
+                    box_threshold=0.35,
+                    text_threshold=0.25,
+                    device="cuda",
+                    debug=True,
+                    output_dir="sam_outputs",
+                )
+                object_mask = np.flipud(obj_masks[-1].cpu().detach().numpy()[0, :, :])
+                frame2 = self.sim.render(camera_name="corner3", width=500, height=500)
+                obj_masks, _, _, pred_phrases, _ = get_seg_mask(
+                    frame2[:, :, ::-1],
+                    self.dino,
+                    self.sam,
+                    text_prompts=["robot, gray nail on wooden box"],
+                    box_threshold=0.4,
+                    text_threshold=0.25,
+                    device="cuda",
+                    debug=True,
+                    output_dir="sam_outputs",
+                )
+                placement_mask = obj_masks[-1].cpu().detach().numpy()[0, :, :]
+                depth_map = get_camera_depth(
+                    camera_name="corner",
+                    camera_width=500,
+                    camera_height=500,
+                    sim=self.sim,
+                )
+                depth_map = np.flipud(np.expand_dims(
+                    CU.get_real_depth_map(sim=self.sim, depth_map=depth_map), -1
+                ))
+
+                # get camera matrices
+                world_to_camera = CU.get_camera_transform_matrix(
+                    sim=self.sim,
+                    camera_name="corner",
+                    camera_height=500,
+                    camera_width=500,
+                )
+                camera_to_world = np.linalg.inv(world_to_camera)
```

Comment:
lots of repetitive code here. i think you can condense all of this stuff by making some helper functions and passing the right args to them.